hash-parser/src/hash.pest

```diff
@@ -2,443 +2,453 @@
-char = @{ "'" ~ (escape | ANY) ~ "'" }
+byte_prefix = _{ "x" }
+unicode_prefix = _{ "u" }
+unicode_grp_left = _{ "{" }
+unicode_grp_right = _{ "}" }
+byte = { byte_prefix ~ hex_digit{2} }
+unicode = { unicode_prefix ~ unicode_grp_left ~ hex_digit{1, 6} ~ unicode_grp_right }
+escape_prefix = _{ "\\" }
+escape = ${ escape_prefix ~ (predefined | byte | unicode) }
 
 /// Bracketed literals
-list_literal  = {"[" ~ ("]" | expr ~ (comma ~ expr)* ~ comma? ~ "]")}
-set_literal   = {"{" ~ (comma ~ &"}" | expr ~ comma ~ &"}" | (expr ~ (comma ~ expr)* ~ comma?)) ~ "}"}
-tuple_literal = {"(" ~ (comma ~ &")" | expr ~ comma ~ &")" | (expr ~ (comma ~ expr)* ~ comma?)) ~ ")"}
+list_literal = { square_left ~ (square_right | expr ~ (comma_op ~ expr)* ~ comma_op? ~ square_right)}
+set_literal = { curly_left ~ (comma_op ~ &curly_right | expr ~ comma_op ~ &curly_right | (expr ~ (comma_op ~ expr)* ~ comma_op?)) ~ curly_right }
+tuple_literal = { paren_left ~ (comma_op ~ &paren_right | expr ~ comma_op ~ &paren_right | (expr ~ (comma_op ~ expr)* ~ comma_op?)) ~ paren_right }
+map_entry = { expr ~ colon_op ~ expr }
+map_literal = { curly_left ~ ((colon_op ~ &curly_right) | (map_entry ~ (comma_op ~ map_entry)* ~ comma_op?)) ~ curly_right }
+
+/// Function literal
+fn_param = { name ~ (colon_op ~ any_type)? }
+fn_params = { paren_left ~ (fn_param ~ (comma_op ~ fn_param)*)? ~ paren_right }
+fn_return_ty = { colon_op ~ any_type }
+fn_body = { arrow_op ~ expr }
+fn_literal = { fn_params ~ fn_return_ty? ~ fn_body }
+
+/// Struct literal
+struct_literal_field = { name ~ eq_op ~ expr }
+struct_literal = { access_name ~ type_args? ~ curly_left ~ (curly_right | struct_literal_field? ~ (comma_op ~ struct_literal_field)* ~ comma_op? ~ curly_right) }
+
+/// ---- Operators ----
+
+/// Assignment operators
+assign_eq_op = { "=" } // Differs from eq_op --- it is not hidden.
+add_eq_op = { "+=" }
+sub_eq_op = { "-=" }
+mul_eq_op = { "*=" }
+div_eq_op = { "/=" }
+mod_eq_op = { "%=" }
+andl_eq_op = { "&&=" }
+orl_eq_op = { "||=" }
+andb_eq_op = { "&=" }
+orb_eq_op = { "|=" }
+xorb_eq_op = { "^=" }
+assign_op = { 
+    assign_eq_op |
+    add_eq_op    |
+    sub_eq_op    |
+    mul_eq_op    |
+    div_eq_op    |
+    mod_eq_op    |
+    andl_eq_op   |
+    orl_eq_op    |
+    andb_eq_op   |
+    orb_eq_op    |
+    xorb_eq_op 
+}
 
-map_entry     = {expr ~ ":" ~ expr}
-map_literal   = {"{" ~ (":" ~ !"}" | (map_entry ~ (comma ~ map_entry)* ~ comma?)) ~ "}"}
+/// Binary operators
+triple_eq_op = { "===" }
+double_eq_op = { "==" }
+neq_op = { "!=" }
+double_neq_op = { "!==" }
+add_op = { "+" }
+sub_op = { "-" }
+mul_op = { "*" }
+div_op = { "/" }
+mod_op = { "%" }
+andl_op = { "&&" }
+orl_op = { "||" }
+shl_op = { "<<" }
+shr_op = { ">>" }
+exp_op = { "**" }
+gt_op = { ">" }
+lt_op = { "<" }
+geq_op = { ">=" }
+leq_op = { "<=" }
+andb_op = { "&" }
+orb_op = { "|" }
+xorb_op = { "^" }
+binary_op = { 
+    triple_eq_op  |
+    double_eq_op  |
+    neq_op        |
+    double_neq_op |
```

Comment:
Ah we didn't think of this before, maybe we should update the hashdocs to list all the operators so we don't forget?